content/4_Gold/BFS.mdx

```diff
@@ -254,6 +254,155 @@ deque.removeLast(); // [1, 3]
+
+For the test input, we start with the following parent array.
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+After visiting children of node $1$:
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td>0</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td>INT_MAX</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+After visiting node $5$ from node $4$
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td>1</td>
+<td>1</td>
+<td>1</td>
+<td className="bg-green-200 dark:bg-green-800">4</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td>1</td>
+<td>1</td>
+<td>1</td>
+<td className="bg-green-200 dark:bg-green-800">2</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+To determine the path, we can backtrack from node $n \rightarrow 1$, in this case $5 \rightarrow 1$, pushing each value that we backtrack into a vector. The path we take is $5 \rightarrow \texttt{parent}[5]=4 \rightarrow \texttt{parent}[4] =1$ which corresponds to the vector $[5, 4, 1]$. We break at node $1$ because it was the initial starting node. Finally, we reverse the vector and print out it's length (in this case, $3$).
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using vi = vector<int>;
+#define pb push_back
+int main() {
+	int N, M; cin >> N >> M;
+	vi dist(N+1,INT_MAX), pre(N+1);
+	vector<vi> adj(N+1);
+	while (M--) {
+		int a,b; cin >> a >> b;
+		adj[a].pb(b), adj[b].pb(a);
+	}
+	queue<int> q;
+	dist[1] = 0; q.push(1);
+	while (!q.empty()) {
+		int x = q.front(); q.pop();
+		for (int t: adj[x]) if (dist[t] == INT_MAX) {
+			dist[t] = dist[x]+1; pre[t] = x;
+			q.push(t);
+		}
+	}
+	if (dist[N] == INT_MAX) cout << "IMPOSSIBLE";
+	else {
+		cout << dist[N]+1 << "\n";
+		vi v{N}; while (v.back() != 1) v.pb(pre[v.back()]);
+		reverse(begin(v), end(v));
+		for (int t: v) cout << t << " ";
+	}
+}
+```
```

Comment:
```suggestion
```cpp
#include <bits/stdc++.h>
using namespace std;
using vi = vector<int>;
#define pb push_back
int main() {
	int N, M; cin >> N >> M;
	vi dist(N+1,INT_MAX), parent(N+1);
	vector<vi> adj(N+1);
	while (M--) {
		int a,b; cin >> a >> b;
		adj[a].pb(b), adj[b].pb(a);
	}
	queue<int> q;
	dist[1] = 0; q.push(1);
	while (!q.empty()) {
		int x = q.front(); q.pop();
		for (int t: adj[x]) if (dist[t] == INT_MAX) {
			dist[t] = dist[x]+1; parent[t] = x;
			q.push(t);
		}
	}
	if (dist[N] == INT_MAX) cout << "IMPOSSIBLE";
	else {
		cout << dist[N]+1 << "\n";
		vi v{N}; while (v.back() != 1) v.pb(parent[v.back()]);
		reverse(begin(v), end(v));
		for (int t: v) cout << t << " ";
	}
}
```
```